cmake/CompilerChecks.cmake

```diff
@@ -23,11 +23,59 @@ if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION)
     endif()
 endif()
 
-# Check for CPUID headers
-try_compile(HAVE_CPUID ${PROJECT_BINARY_DIR}
-            ${PROJECT_SOURCE_DIR}/cmake/check_cpuid.cpp)
-if(HAVE_CPUID)
-    message(STATUS "CPUID header found -- using hardware math acceleration when available")
+# Check for SIMD headers
+include(CheckIncludeFile)
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+    CHECK_INCLUDE_FILE("cpuid.h" HAVE_CPUID)
 else()
-    message(STATUS "CPUID header not found -- using software math")
+    CHECK_INCLUDE_FILE("intrin.h" HAVE_CPUID)
 endif()
+
+CHECK_INCLUDE_FILE("immintrin.h" HS_AVX2)
+CHECK_INCLUDE_FILE("immintrin.h" HS_AVX)
+CHECK_INCLUDE_FILE("nmmintrin.h" HS_SSE42)
+CHECK_INCLUDE_FILE("smmintrin.h" HS_SSE41)
+CHECK_INCLUDE_FILE("tmmintrin.h" HS_SSE4)
+CHECK_INCLUDE_FILE("pmmintrin.h" HS_SSE3)
+CHECK_INCLUDE_FILE("emmintrin.h" HS_SSE2)
+CHECK_INCLUDE_FILE("xmmintrin.h" HS_SSE1)
+
+# GCC requires us to set the -m<instructionset> flag for the source file using the intrinsics.
+# We can't do that project-wide or we'll just crash on launch with an illegal instruction on some
+# systems. So, we have another helper method...
+function(plasma_target_simd_sources TARGET)
+    set(_INSTRUCTION_SETS "SSE1;SSE2;SSE3;SSE4;SSE41;SSE42;AVX;AVX2")
+    set(_GCC_ARGS "-msse;-msse2;-msse3;-msse4;-msse4.1;-msse4.2;-mavx;-mavx2")
+    cmake_parse_arguments(_passf "" "SOURCE_GROUP" "${_INSTRUCTION_SETS}" ${ARGN})
+
+    # Hack: if we ever bump to CMake 3.17, use ZIP_LISTS.
+    list(LENGTH _INSTRUCTION_SETS _LENGTH)
+    math(EXPR _LENGTH "${_LENGTH} - 1")
+
+    foreach(i RANGE ${_LENGTH})
+        # foreach(i IN ZIP_LISTS _INSTRUCTION_SETS _GCC_ARGS)
+        list(GET _INSTRUCTION_SETS ${i} i_0)
+        list(GET _GCC_ARGS ${i} i_1)
```

Comment:
Even if you are using ZIP_LISTS, I would prefer to see reasonable variable names, rather than `${i_0}` and `${i_1}`

Reply:
Sadly, we aren't using `ZIP_LISTS`, so reasonable variable names are definitely an option.